src/H5Cimage.c

```diff
@@ -2546,6 +2590,50 @@ H5C__reconstruct_cache_contents(H5F_t *f, H5C_t *cache_ptr)
     } /* end if */
 
 done:
+    if (FAIL == ret_value && pf_entry_ptr) {
+#ifndef NDEBUG
+        /* If we failed during reconstruction, remove reconstructed entries */
+        H5C_recon_entry_t *recon_entry, *tmp;
+
+        HASH_ITER(hh, recon_table, recon_entry, tmp)
+        {
+            H5C_cache_entry_t *entry_ptr = recon_entry->entry_ptr;
+
+            /* Only touch entries from the image reconstruction */
+            if (entry_ptr->type->id == H5AC_PREFETCHED_ENTRY_ID) {
+                /*                 if (!entry_ptr->is_pinned && !entry_ptr->is_protected)
+                 */
+                H5AC_expunge_entry(f, H5AC_PREFETCHED_ENTRY, recon_entry->addr, H5AC__NO_FLAGS_SET);
+
+                if (entry_ptr->image_ptr)
+                    H5MM_xfree(entry_ptr->image_ptr);
+            }
+
+            HASH_DEL(recon_table, recon_entry);
+            H5MM_xfree(recon_entry);
+        }
+        /* The temporary hash table should be empty */
+        assert(recon_table == NULL);
+#endif /* NDEBUG */
+
+        /* Free the half-processed entry */
+        if (pf_entry_ptr->image_ptr)
```

Comment:
You will of course need an `if (pf_entry_ptr) {` here if you follow my above suggestion. And I think a `pf_entry_ptr = NULL;` at the end of the main "reconstruct entries in image" for loop